File: Model/GraphMaster.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
using System.Windows.Forms;

namespace Isocline
{
    class GraphMaster
    {
        private double min = -10;
        private double max = 10;
        private int width;
        private int height;
        private static Font font = new Font("Times New Roman", 14);
        private PictureBox picture;
        public GraphMaster(double _min, double _max)
        {
            min = _min;
            max = _max;
        }

        public Image CreateGraph(int w, int h, Action<Graphics> drawing)
        {
            width = w;
            height = h;
            var bmp = new Bitmap(w, h);
            using (var g = Graphics.FromImage(bmp))
            {
                g.FillRectangle(Brushes.White, 0, 0, w, h);
                drawing(g);
                DrawCoord(g);
            }
            return bmp;
        }

        private void DrawCoord(Graphics g)
        {
            g.DrawLine(Pens.Black, Helpers.Transform(0, min,min,max,width,height), Helpers.Transform(0, max, min, max, width, height));
            g.DrawLine(Pens.Black, Helpers.Transform(min, 0, min, max, width, height), Helpers.Transform(max, 0, min, max, width, height));
            for (var i = min; i < max; i += 1)
            {
                var transformPointY = Helpers.Transform(0, i, min, max, width, height);
                g.DrawString(String.Format("{0}", Math.Round(i, 2)), font, Brushes.Black, transformPointY);
                g.DrawLine(Pens.Black, transformPointY.X - 3, transformPointY.Y, transformPointY.X + 3, transformPointY.Y);

                var transformPointX = Helpers.Transform(i, 0, min, max, width, height);
                g.DrawString(String.Format("{0}", Math.Round(i, 2)), font, Brushes.Black, transformPointX);
                g.DrawLine(Pens.Black, transformPointX.X, transformPointX.Y - 3, transformPointX.X, transformPointX.Y + 3);
            }
        }

        public void DrawQuadraticInterpolation(Graphics g, Func<double, double, double> dy, double x0, double y0)
        {
            double x = x0, y = y0, step = 0.001;

            // Iterate forward
            while (x <= max)
            {
                var k1 = dy(x, y);
                var k2 = dy(x + step / 2, y + step / 2 * k1);
                var k3 = dy(x + step, y + step * k2);

                var nx = x + step;
                var ny = y + step / 6 * (k1 + 4 * k2 + k3);

                try
                {
                    g.DrawLine(Pens.Black, Helpers.Transform(x, y,min, max, width, height), Helpers.Transform(nx, ny, min, max, width, height));
                }
                catch { }

                x = nx;
                y = ny;
            }

            x = x0;
            y = y0;

            // Iterate backward
            while (x >= min)
            {
                var k1 = dy(x, y);
                var k2 = dy(x - step / 2, y - step / 2 * k1);
                var k3 = dy(x - step, y - step * k2);

                var nx = x - step;
                var ny = y - step / 6 * (k1 + 4 * k2 + k3);

                try
                {
                    g.DrawLine(Pens.Black, Helpers.Transform(x, y, min, max, width, height), Helpers.Transform(nx, ny, min, max, width, height));
                }
                catch { }

                x = nx;
                y = ny;
            }
        }
        public void DrawLinearInterpolation(Graphics g, Func<double, double, double> dy, double x0, double y0)
        {
            double x = x0, y = y0, step = 0.001;
            while (x <= max)
            {
                var k = dy(x, y);
                var nx = x + step;
                var ny = y + step * k;

                try
                {
                    g.DrawLine(Pens.Black, Helpers.Transform(x, y, min, max, width, height), Helpers.Transform(nx, ny, min, max, width, height));
                }
                catch { };

                x = nx;
                y = ny;
            }

            x = x0;
            y = y0;

            while (x >= min)
            {
                var k = dy(x, y);
                var nx = x - step;
                var ny = y - step * k;

                try
                {
                    g.DrawLine(Pens.Black, Helpers.Transform(x, y, min, max, width, height), Helpers.Transform(nx, ny, min, max, width, height));
                }
                catch { };

                x = nx;
                y = ny;
            }
        }
    }
}

File: Model/GraphMaster.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Namespace Isocline
    Friend Class GraphMaster
        Private min As Double = -10
        Private max As Double = 10
        Private width As Integer
        Private height As Integer
        Private Shared font As Font = New Font("Times New Roman", 14)
        Private picture As PictureBox
        Public Sub New(ByVal _min As Double, ByVal _max As Double)
            min = _min
            max = _max
        End Sub

        Public Function CreateGraph(ByVal w As Integer, ByVal h As Integer, ByVal drawing As Action(Of Graphics)) As Image
            width = w
            height = h
            Dim bmp = New Bitmap(w, h)
            Using g = Graphics.FromImage(bmp)
                g.FillRectangle(Brushes.White, 0, 0, w, h)
                drawing(g)
                DrawCoord(g)
            End Using
            Return bmp
        End Function

        Private Sub DrawCoord(ByVal g As Graphics)
            g.DrawLine(Pens.Black, Transform(0, min), Transform(0, max))
            g.DrawLine(Pens.Black, Transform(min, 0), Transform(max, 0))
            Dim [step] = (max - min) / 20
            Dim i = min

            While i < max
                g.DrawString(String.Format("{0}", Math.Round(i, 2)), font, Brushes.Black, Transform(0, i))
                g.DrawString(String.Format("{0}", Math.Round(i, 2)), font, Brushes.Black, Transform(i, 0))
                i += [step]
            End While
        End Sub

        Public Function Transform(ByVal x As Double, ByVal y As Double) As Point

            Return New Point((x + (max - min) / 2) / (max - min) * width, (-y + (max - min) / 2) / (max - min) * height)
        End Function

        Public Sub DrawSolution(ByVal g As Graphics, ByVal dy As Func(Of Double, Double, Double), ByVal x0 As Double, ByVal y0 As Double)
            Dim x = x0, y = y0
            Dim [step] = 0.001 ' Adjust the step size as needed

            While x <= max
                Dim k1 = [step] * dy(x, y)
                Dim k2 = [step] * dy(x + 0.5 * [step], y + 0.5 * k1)
                Dim k3 = [step] * dy(x + 0.5 * [step], y + 0.5 * k2)
                Dim k4 = [step] * dy(x + [step], y + k3)

                Dim nx = x + [step]
                Dim ny = y + (k1 + 2 * k2 + 2 * k3 + k4) / 6

                Try
                    g.DrawLine(Pens.Black, Transform(x, y), Transform(nx, ny))
                Catch
                End Try

                x = nx
                y = ny
            End While

            x = x0
            y = y0

            While x >= min
                Dim k1 = [step] * dy(x, y)
                Dim k2 = [step] * dy(x - 0.5 * [step], y - 0.5 * k1)
                Dim k3 = [step] * dy(x - 0.5 * [step], y - 0.5 * k2)
                Dim k4 = [step] * dy(x - [step], y - k3)

                Dim nx = x - [step]
                Dim ny = y - (k1 + 2 * k2 + 2 * k3 + k4) / 6

                Try
                    g.DrawLine(Pens.Black, Transform(x, y), Transform(nx, ny))
                Catch
                End Try

                x = nx
                y = ny
            End While
        End Sub

    End Class
End Namespace
